################################################################################
SHORT DESCRIPTION: 
################################################################################
Introduced option to disable VMCS shadowing.

################################################################################
LONG DESCRIPTION: 
################################################################################
This patch was created to support vSentry running as a nested VMM on OpenXT.
It is not entirely clear why VMCS shadowing needs to be disabled since it is a
feature to assist nested VMMs. It may not be needed any longer to support
vSentry but it will be left here as a placeholder since it does nothing by
default.

################################################################################
CHANGELOG 
################################################################################

################################################################################
REMOVAL 
################################################################################
If no longer needed for vSentry interop, then it can be removed.

################################################################################
UPSTREAM PLAN
################################################################################

################################################################################
INTERNAL DEPENDENCIES 
################################################################################

################################################################################
PATCHES 
################################################################################
--- a/xen/arch/x86/hvm/vmx/vmcs.c
+++ b/xen/arch/x86/hvm/vmx/vmcs.c
@@ -51,6 +51,9 @@ boolean_param("unrestricted_guest", opt_
 static bool_t __read_mostly opt_apicv_enabled = 1;
 boolean_param("apicv", opt_apicv_enabled);
 
+static bool_t __read_mostly opt_vmcs_shadowing = 1;
+boolean_param("vmcs_shadowing", opt_vmcs_shadowing);
+
 /*
  * These two parameters are used to config the controls for Pause-Loop Exiting:
  * ple_gap:    upper bound on the amount of time between two successive
@@ -371,6 +374,11 @@ static int vmx_init_vmcs_config(void)
     if ( !(_vmx_secondary_exec_control & SECONDARY_EXEC_ENABLE_VM_FUNCTIONS) )
         _vmx_secondary_exec_control &= ~SECONDARY_EXEC_ENABLE_VIRT_EXCEPTIONS;
 
+    if ( !opt_vmcs_shadowing ) {
+        _vmx_secondary_exec_control &= ~SECONDARY_EXEC_ENABLE_VMCS_SHADOWING;
+        printk(XENLOG_INFO "Disable VMCS Shadowing.\n");
+    }
+
     min = 0;
     opt = (VM_ENTRY_LOAD_GUEST_PAT | VM_ENTRY_LOAD_GUEST_EFER |
            VM_ENTRY_LOAD_BNDCFGS);
